src/main/java/io/quarkus/search/app/fetching/FetchingService.java

```diff
@@ -69,23 +73,54 @@ public QuarkusIO fetchQuarkusIo() {
+    private GitCloneDirectory fetchQuarkusIoSite(String siteName, URI gitUri, Branches branches,
             CloseableDirectory unzipDir) {
+        // We only want to clean up a clone directory if it is a clone from a remote repository or an unzipped one.
+        //  If we got a local repository we want to keep it as is and no cloning is needed.
+        //  If it was a local unzipped version -- the "cloneDir" will be removed sooner than we are done with indexing,
+        //  so we want to "clone" that directory, and then it'll be removed as if a remote repository clone.
+        boolean requiresCloning = !isFile(gitUri) || isZip(gitUri);
+        URI requestedGitUri = gitUri;
+        CloseableDirectory cloneDir = null;
         try {
+            GitCloneDirectory.GitDirectoryDetails repository = repositories.get(gitUri);
+
             if (LaunchMode.DEVELOPMENT.equals(LaunchMode.current()) && isZip(gitUri)) {
+                if (repository != null) {
+                    // We are working with a zip file, so we have nothing to refresh as there's no actual remote available;
+                    //   just return the same repository without any changes:
+                    return repository.open();
+                }
+
                 Log.warnf("Unzipping '%s': this application is most likely indexing only a sample of %s."
                         + " See README to index the full website.",
                         gitUri, siteName);
                 Path unzippedPath = unzipDir.path().resolve(siteName);
                 unzip(Path.of(gitUri), unzippedPath);
                 gitUri = unzippedPath.toUri();
                 // Fall-through and clone the directory.
-                // While technically unnecessary (we could use the unzipped directory directly),
-                // this cloning ensures we run the same code in dev mode as in prod.
+                // This cloning is as if we are cloning a remote repository.
+            } else {
+                if (repository != null) {
+                    // It's not a zip, so it may be either a local directory or a remote repository
+                    // let's pull the changes from it as it shouldn't cause any exceptions, as a remote is actually out there.
+                    return repository.pull(branches);
+                }
             }
-            return gitClone(siteName, gitUri, List.of(sourceBranch, pagesBranch),
-                    (git, directory) -> new GitCloneDirectory(git, directory, pagesBranch));
+
+            cloneDir = requiresCloning
+                    ? CloseableDirectory.temp(siteName)
+                    : CloseableDirectory.of(Paths.get(gitUri));
+
+            closeableDirectories.add(cloneDir);
+
+            repository = new GitCloneDirectory.GitDirectoryDetails(cloneDir.path(), branches.pages());
+            repositories.put(requestedGitUri, repository);
+
+            // If we have a local repository -- just open it, clone it otherwise:
```

Comment:
I think if you have a local repository, you should probably pull? That is, if you're trying to solve #126. And I think you are?

I suspect this could solve some of the problems I've been experiencing with this patch... I'm getting things like this:

```
11:47:15 WARN  [c.ar.at.arjuna] ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffffc0a8012f:a64b:659e75a6:4fb, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@79a6c5cc >: org.hibernate.search.util.common.SearchException: HSEARCH700083: Exception while building document for entity 'Guide#http://localhost:4000/guides/grpc-generation-reference': Failed to read 'GitInputProvider{path='guides/grpc-generation-reference.html'}' for indexing: Missing file 'guides/grpc-generation-reference.html' in 'tree 86751bb1d0372a2e7c9f6714a13a338367ee0474 ------p'
        at org.hibernate.search.mapper.pojo.work.impl.PojoDocumentContributor.contribute(PojoDocumentContributor.java:62)
        at org.hibernate.search.backend.elasticsearch.index.impl.ElasticsearchIndexManagerImpl.createDocument(ElasticsearchIndexManagerImpl.java:164)
        at org.hibernate.search.backend.elasticsearch.work.execution.impl.ElasticsearchIndexIndexingPlan.index(ElasticsearchIndexIndexingPlan.java:104)
        at org.hibernate.search.backend.elasticsearch.work.execution.impl.ElasticsearchIndexIndexingPlan.add(ElasticsearchIndexIndexingPlan.java:54)
        at org.hibernate.search.mapper.pojo.work.impl.PojoTypeIndexingPlanIndexDelegate.add(PojoTypeIndexingPlanIndexDelegate.java:61)
        at org.hibernate.search.mapper.pojo.work.impl.AbstractPojoTypeIndexingPlan$AbstractEntityState.delegateAdd(AbstractPojoTypeIndexingPlan.java:448)
        at org.hibernate.search.mapper.pojo.work.impl.AbstractPojoTypeIndexingPlan$AbstractEntityState.sendCommandsToDelegate(AbstractPojoTypeIndexingPlan.java:408)
        at org.hibernate.search.mapper.pojo.work.impl.AbstractPojoTypeIndexingPlan.process(AbstractPojoTypeIndexingPlan.java:160)
        at org.hibernate.search.mapper.pojo.work.impl.PojoIndexingPlanImpl.process(PojoIndexingPlanImpl.java:171)
        at org.hibernate.search.mapper.orm.event.impl.HibernateSearchEventListener.onFlush(HibernateSearchEventListener.java:241)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
        at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
        at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:485)
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2301)
        at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1966)
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
        at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
        at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
        at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:351)
        at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:69)
        at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:138)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1271)
        at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
        at io.quarkus.narayana.jta.runtime.NotifyingUserTransaction.commit(NotifyingUserTransaction.java:34)
        at jakarta.transaction.NarayanaJtaProducers_ProducerMethod_userTransaction_7zPXeUWVelZ-AzCmqVtVVw94OtA_ClientProxy.commit(Unknown Source)
        at io.quarkus.narayana.jta.QuarkusTransactionImpl.callInOurTx(QuarkusTransactionImpl.java:154)
        at io.quarkus.narayana.jta.QuarkusTransactionImpl.callRequireNew(QuarkusTransactionImpl.java:106)
        at io.quarkus.narayana.jta.QuarkusTransactionImpl.call(QuarkusTransactionImpl.java:29)
        at io.quarkus.narayana.jta.TransactionRunnerImpl.run(TransactionRunnerImpl.java:26)
        at io.quarkus.search.app.indexing.IndexingService.indexBatch(IndexingService.java:250)
        at io.quarkus.search.app.indexing.IndexingService_Subclass.indexBatch$$superforward(Unknown Source)
        at io.quarkus.search.app.indexing.IndexingService_Subclass$$function$$1.apply(Unknown Source)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
        at io.quarkus.arc.impl.ActivateRequestContextInterceptor.invoke(ActivateRequestContextInterceptor.java:129)
        at io.quarkus.arc.impl.ActivateRequestContextInterceptor.aroundInvoke(ActivateRequestContextInterceptor.java:33)
        at io.quarkus.arc.impl.ActivateRequestContextInterceptor_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
        at io.quarkus.search.app.indexing.IndexingService_Subclass.indexBatch(Unknown Source)
        at io.quarkus.search.app.indexing.IndexingService.lambda$indexAll$9(IndexingService.java:238)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalStateException: Failed to read 'GitInputProvider{path='guides/grpc-generation-reference.html'}' for indexing: Missing file 'guides/grpc-generation-reference.html' in 'tree 86751bb1d0372a2e7c9f6714a13a338367ee0474 ------p'
        at io.quarkus.search.app.hibernate.InputProviderHtmlBodyTextBridge.toIndexedValue(InputProviderHtmlBodyTextBridge.java:47)
        at io.quarkus.search.app.hibernate.InputProviderHtmlBodyTextBridge.toIndexedValue(InputProviderHtmlBodyTextBridge.java:16)
        at io.quarkus.search.app.hibernate.I18nFullTextFieldAnnotationProcessor$LanguageAlternativeBinderDelegate$Bridge.write(I18nFullTextFieldAnnotationProcessor.java:106)
        at io.quarkus.search.app.hibernate.I18nFullTextFieldAnnotationProcessor$LanguageAlternativeBinderDelegate$Bridge.write(I18nFullTextFieldAnnotationProcessor.java:94)
        at org.hibernate.search.mapper.pojo.bridge.builtin.alternative.impl.AlternativeBinderImpl$Bridge.write(AlternativeBinderImpl.java:111)
        at org.hibernate.search.mapper.pojo.processing.impl.PojoIndexingProcessorTypeBridgeNode.process(PojoIndexingProcessorTypeBridgeNode.java:45)
        at org.hibernate.search.mapper.pojo.processing.impl.PojoIndexingProcessorMultiNode.process(PojoIndexingProcessorMultiNode.java:45)
        at org.hibernate.search.mapper.pojo.processing.impl.PojoIndexingProcessorOriginalTypeNode.process(PojoIndexingProcessorOriginalTypeNode.java:63)
        at org.hibernate.search.mapper.pojo.work.impl.PojoDocumentContributor.contribute(PojoDocumentContributor.java:55)
        ... 47 more
Caused by: java.lang.IllegalStateException: Missing file 'guides/grpc-generation-reference.html' in 'tree 86751bb1d0372a2e7c9f6714a13a338367ee0474 ------p'
        at io.quarkus.search.app.util.GitUtils.file(GitUtils.java:41)
        at io.quarkus.search.app.util.GitInputProvider.open(GitInputProvider.java:26)
        at io.quarkus.search.app.hibernate.InputProviderHtmlBodyTextBridge.toIndexedValue(InputProviderHtmlBodyTextBridge.java:19)
        ... 55 more
```

Reply:
Yeah, I've seen you creating that one, and thought I'll put something in that would make it easier to address the next task since I'm changing the code already.
🙈 🙈 🙈 🙈 🙈 
See the `Make sure correct refs are inspected for pages tree` commit for this particular problem.... 
since we were looking at `origin/bla-bla-bla` and if a local repo has pulled the changes from an upstream but didn't push that to `origin` ... we were ending up with this error. 
and then once I've noticed it ... I just added that commit and also a branch checkout as well 🙈

Reply:
Ok...

I just tried with your latest changes and I'm still getting what appears to be the same error (see below).

May this be related to the fact that I have multiple remotes on my clone, and you just pick the first one in your code (which may not be the right one)?

```
❯ git remote -v
contrib-marko-bekhta    git@github.com:marko-bekhta/quarkusio.github.io.git (fetch)
contrib-marko-bekhta    git@github.com:marko-bekhta/quarkusio.github.io.git (push)
origin  git@github.com:yrodiere/quarkusio.github.io.git (fetch)
origin  git@github.com:yrodiere/quarkusio.github.io.git (push)
upstream        git@github.com:quarkusio/quarkusio.github.io.git (fetch)
upstream        git@github.com:quarkusio/quarkusio.github.io.git (push)
```

Maybe you should have a list of common remote names to try first, e.g. `upstream`, then `origin`, then whatever the first remote is?

I think we do that in `QuarkusIOSample`: https://github.com/quarkusio/search.quarkus.io/blob/55749d39e810965c8ab92cddaebc98d1bba0affd/src/test/java/io/quarkus/search/app/testsupport/QuarkusIOSample.java#L85 Maybe we need to use similar code here...

Reply:
hmm yeah, let me change this part then 😃 and thanks for testing it!